src/package.js

```diff
@@ -0,0 +1,1106 @@
+      } else {
+        this.compatible = true
+      }
+    }
+    return this.compatible
+  }
+
+  // Extended: Rebuild native modules in this package's dependencies for the
+  // current version of Atom.
+  //
+  // Returns a {Promise} that resolves with an object containing `code`,
+  // `stdout`, and `stderr` properties based on the results of running
+  // `apm rebuild` on the package.
+  rebuild () {
+    return new Promise(resolve =>
+      this.runRebuildProcess(result => {
+        if (result.code === 0) {
+          global.localStorage.removeItem(this.getBuildFailureOutputStorageKey())
+        } else {
+          this.compatible = false
+          global.localStorage.setItem(this.getBuildFailureOutputStorageKey(), result.stderr)
+        }
+        global.localStorage.setItem(this.getIncompatibleNativeModulesStorageKey(), '[]')
+        resolve(result)
+      })
+    )
+  }
+
+  // Extended: If a previous rebuild failed, get the contents of stderr.
+  //
+  // Returns a {String} or null if no previous build failure occurred.
+  getBuildFailureOutput () {
+    return global.localStorage.getItem(this.getBuildFailureOutputStorageKey())
+  }
+
+  runRebuildProcess (done) {
+    let stderr = ''
+    let stdout = ''
+    return new BufferedProcess({
+      command: this.packageManager.getApmPath(),
+      args: ['rebuild', '--no-color'],
+      options: {cwd: this.path},
+      stderr (output) { stderr += output },
+      stdout (output) { stdout += output },
+      exit (code) { done({code, stdout, stderr}) }
+    })
+  }
+
+  getBuildFailureOutputStorageKey () {
+    return `installed-packages:${this.name}:${this.metadata.version}:build-error`
+  }
+
+  getIncompatibleNativeModulesStorageKey () {
+    const electronVersion = process.versions.electron
+    return `installed-packages:${this.name}:${this.metadata.version}:electron-${electronVersion}:incompatible-native-modules`
+  }
+
+  getCanDeferMainModuleRequireStorageKey () {
+    return `installed-packages:${this.name}:${this.metadata.version}:can-defer-main-module-require`
+  }
+
+  // Get the incompatible native modules that this package depends on.
+  // This recurses through all dependencies and requires all modules that
+  // contain a `.node` file.
+  //
+  // This information is cached in local storage on a per package/version basis
+  // to minimize the impact on startup time.
+  getIncompatibleNativeModules () {
+    if (!this.packageManager.devMode) {
+      try {
+        const arrayAsString = global.localStorage.getItem(this.getIncompatibleNativeModulesStorageKey())
+        if (arrayAsString) return JSON.parse(arrayAsString)
+      } catch (error1) {}
+    }
+
+    const incompatibleNativeModules = []
+    for (let nativeModulePath of this.getNativeModuleDependencyPaths()) {
+      try {
+        require(nativeModulePath)
+      } catch (error) {
+        let version
+        try {
+          ({version} = require(`${nativeModulePath}/package.json`))
```

Comment:
are parens doing anything here?

Reply:
Yes, you cannot have a bracket start a line because it interprets it as a block.